File: Cracking the Coding Interview/8. Recursion and Dynamic Programming/8.6 towerOfHanoi.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;

void towerHanoi(int platesRemaining , array<stack <int> , 3> &pegs, int from_peg,int to_peg,int using_peg);

void makeHanoi(int NumOfElemets){

    array<stack <int> ,3> pegs;

    for(int i=NumOfElemets;i>0;i--){
        pegs[0].push(i);
    }

    towerHanoi(NumOfElemets,pegs,0,1,2);

}

void towerHanoi(int platesRemaining , array<stack <int> , 3> &pegs, int from_peg,int to_peg,int using_peg){

    if(platesRemaining>0){
        towerHanoi(platesRemaining-1,pegs,from_peg,using_peg,to_peg);
        
        pegs[to_peg].push(pegs[from_peg].top());

        int poped = pegs[from_peg].top();
        pegs[from_peg].pop();
        cout<<"Moved " << poped<<" from "<<from_peg << " to the "<<to_peg<<endl;

        towerHanoi(platesRemaining-1,pegs,using_peg,to_peg,from_peg);
    }
}

int main(){

    makeHanoi(3);
    return 0;
}

File: Cracking the Coding Interview/8. Recursion and Dynamic Programming/8.6 towerOfHanoi.py
from collections import deque


pegs=[]
pegs.append(deque()) 
pegs.append(deque())
pegs.append(deque())

def init_pegs(numPlates):
    for i in range(numPlates,0,-1):
        pegs[0].append(i)



def towerOfHanoi(numPlates,pegs,from_peg,to_peg,using_peg):
    if(numPlates>0):
        towerOfHanoi(numPlates-1,pegs,from_peg,using_peg,to_peg)

        pegs[to_peg].append(pegs[from_peg].pop())
        print('Moved from ',from_peg,' to ',to_peg)
        towerOfHanoi(numPlates-1,pegs,using_peg,to_peg,from_peg)

init_pegs(3)
towerOfHanoi(3,pegs,0,1,2)